msi/createmsi.py

```diff
@@ -78,78 +35,177 @@ def __init__(self, dirs, files):
+        import xml.dom.minidom
+        doc = xml.dom.minidom.parse(self.main_xml)
+        with open(self.main_xml, 'w') as of:
+            of.write(doc.toprettyxml())
+
+    def build_features(self, nodes, top_feature, staging_dir):
+        feature = ET.SubElement(top_feature, 'Feature',  self.feature_properties[staging_dir])
+        for component_id in self.feature_components[staging_dir]:
+            ET.SubElement(feature, 'ComponentRef', {
+                'Id': component_id,
+                })
+        
+
+    def create_xml(self, nodes, current_dir, parent_xml_node, staging_dir):
+        cur_node = nodes[current_dir]
         if cur_node.files:
-            ofile.write("<Component Id='ApplicationFiles%d' Guid='%s' %s>\n" % (self.component_num, gen_guid(), self.component_platform))
+            component_id = 'ApplicationFiles%d' % self.component_num
+            comp_xml_node = ET.SubElement(parent_xml_node, 'Component', {
+                'Id': component_id,
+                'Guid': gen_guid(),
+                })
+            self.feature_components[staging_dir].append(component_id)
+            if self.bytesize == 64:
+                comp_xml_node.set('Win64', 'yes')
             if self.component_num == 0:
-                ofile.write(path_addition_xml)
+                ET.SubElement(comp_xml_node, 'Environment', {
+                    'Id': 'Environment',
+                    'Name': 'PATH',
+                    'Part': 'last',
+                    'System': 'yes',
+                    'Action': 'set',
+                    'Value': '[INSTALLDIR]',
+                })
             self.component_num += 1
             for f in cur_node.files:
-                file_source = os.path.join(root, f).replace('\\', '\\\\')
-                file_id = os.path.join(root, f).replace('\\', '_').replace('#', '_').replace('-', '_')
-                ofile.write(file_templ % (file_id, f, file_source))
-            ofile.write('</Component>\n')
+                file_source = os.path.join(current_dir, f).replace('\\', '\\\\')
+                file_id = os.path.join(current_dir, f).replace('\\', '_').replace('#', '_').replace('-', '_')
+                ET.SubElement(comp_xml_node, 'File', {
+                    'Id': file_id,
+                    'Name': f,
+                    'Source': os.path.join(current_dir, f),
+                    })
 
         for dirname in cur_node.dirs:
-            dir_id = os.path.join(root, dirname).replace('\\', '_').replace('/', '_')
-            ofile.write('''<Directory Id="%s" Name="%s">\n''' % (dir_id, dirname))
-            self.create_xml(nodes, ofile, os.path.join(root, dirname))
-            ofile.write('</Directory>\n')
-
+            dir_id = os.path.join(current_dir, dirname).replace('\\', '_').replace('/', '_')
+            dir_node = ET.SubElement(parent_xml_node, 'Directory', {
+                'Id': dir_id,
+                'Name': dirname,
+                })
```

Comment:
__[flake8]__
* [E123] closing bracket does not match indentation of opening bracket's line

<!-- Comment by [SideCI](https://sideci.com) -->